src/mod_mam_elasticsearch_arch.erl

```diff
@@ -0,0 +1,288 @@
+%%------------------------------------------------------------------
+%% Copyright 2018 Erlang Solutions Ltd.
+%%
+%% Licensed under the Apache License, Version 2.0 (the "License");
+%% you may not use this file except in compliance with the License.
+%% You may obtain a copy of the License at
+%%
+%% http://www.apache.org/licenses/LICENSE-2.0
+%%
+%% Unless required by applicable law or agreed to in writing, software
+%% distributed under the License is distributed on an "AS IS" BASIS,
+%% WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+%% See the License for the specific language governing permissions and
+%% limitations under the License.
+%%
+%% @doc ElasticSearch backend for Message Archive Management for
+%% one-to-one messages.
+%%%------------------------------------------------------------------
+-module(mod_mam_elasticsearch_arch).
+
+-behaviour(gen_mod).
+-behaviour(ejabberd_gen_mam_archive).
+
+%% gen_mod callbacks
+-export([start/2]).
+-export([stop/1]).
+
+%% ejabberd_gen_mam_archive callbacks
+-export([archive_message/9]).
+-export([lookup_messages/3]).
+-export([remove_archive/4]).
+-export([archive_size/4]).
+
+-include("mongoose.hrl").
+-include("mongoose_rsm.hrl").
+-include("mod_mam.hrl").
+-include("jlib.hrl").
+
+-define(INDEX_NAME, <<"messages">>).
+-define(TYPE_NAME, <<"pm">>).
+
+%%-------------------------------------------------------------------
+%% gen_mod callbacks
+%%-------------------------------------------------------------------
+
+-spec start(jid:server(), list()) -> ok.
+start(Host, _Opts) ->
+    ejabberd_hooks:add(hooks(Host)),
+    ok.
+
+-spec stop(jid:server()) -> ok.
+stop(Host) ->
+    ejabberd_hooks:delete(hooks(Host)),
+    ok.
+
+%%-------------------------------------------------------------------
+%% ejabberd_gen_mam_archive callbacks
+%%-------------------------------------------------------------------
+
+archive_message(_Result, Host, MessageId, _UserId, LocalJid, RemoteJid, SourceJid, _Dir, Packet) ->
+    Owner = mod_mam_utils:bare_jid(LocalJid),
+    Remote = mod_mam_utils:bare_jid(RemoteJid),
+    SourceBinJid = mod_mam_utils:full_jid(SourceJid),
+    DocId = make_document_id(Owner, MessageId),
+    Doc = make_document(MessageId, Owner, Remote, SourceBinJid, Packet),
+    case mongoose_elasticsearch:insert_document(?INDEX_NAME, ?TYPE_NAME, DocId, Doc) of
+        {ok, _} ->
+            ok;
+        {error, _} = Err ->
+            ?ERROR_MSG("Failed to archive message in ElasticSearch: ~p", [Err]),
+            ejabberd_hooks:run(mam_drop_message, Host, [Host]),
+            Err
+    end.
+
+lookup_messages(_Result, _Host, Params) ->
+    SearchQuery0 = build_search_query(Params),
```

Comment:
SearchQuery instead of SearchQuery0

Reply:
why ?

Reply:
Hmm, because this code 
```erlang
Req1 = escalus_stanza:to(Req, To),
State1 = set_opts(DefRoomOpts, State),
```
is slightly more popular than

```erlang
State1 = State0#state{override_global = true,
```